src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,128 @@
+        {
+            private readonly DocumentId _documentId;
+            private readonly string _documentName;
+            private readonly IReadOnlyList<string> _documentFolders;
+            private readonly ProjectId _projectId;
+            private readonly OptionSet _optionSet;
+
+            internal RenameDocumentActionSet(
+                ImmutableArray<RenameDocumentAction> actions,
+                ProjectId projectId,
+                DocumentId documentId,
+                string documentName,
+                IReadOnlyList<string> documentFolders,
+                OptionSet optionSet)
+            {
+                ApplicableActions = actions;
+                _documentFolders = documentFolders;
+                _documentId = documentId;
+                _documentName = documentName;
+                _projectId = projectId;
+                _optionSet = optionSet;
+            }
+
+            /// <summary>
+            /// All applicable actions computed for the action
+            /// </summary>
+            public ImmutableArray<RenameDocumentAction> ApplicableActions { get; }
+
+            /// <summary>
+            /// Same as calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/> with 
+            /// <see cref="ApplicableActions"/> as the argument
+            /// </summary>
+            public Task<Solution> UpdateSolutionAsync(Solution solution, CancellationToken cancellationToken)
+                => UpdateSolutionAsync(solution, ApplicableActions, cancellationToken);
+
+            /// <summary>
+            /// Applies each <see cref="RenameDocumentAction"/> in order and returns the final solution. 
+            /// All actions must be contained in <see cref="ApplicableActions" />
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (solution is null)
+                {
+                    throw new ArgumentNullException(nameof(solution));
+                }
+
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new ArgumentException(WorkspacesResources.Cannot_apply_action_that_is_not_in_applicableactions);
+                }
+
+                // Prior to updating the solution it's possible the document id has changed between the time 
+                // the document action info was generated and when the solution update is applied. We
+                // do a best effort to still locate the document if the id has changed.
+                var document = GetDocument(solution);
```

Comment:
is this scenario really important to support?  why don't we just say: if the solution has changed that much, the action is no longer applicable.

For that matter... could the action-set itself just hold onto the starting solution, and then just give you teh result of applying things on that known good solution?  the idea of producing actions against one snapshot to eventually run against another snapshot is a bit... strange (and def not something we really see/do elsewhere in Roslyn).

Reply:
this comment still applies.

Reply:
The reason for this logic will be discussed in https://github.com/dotnet/roslyn/pull/41267/files#r412654520